File: 2024/day07/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

sub parse_input {
    my ($text) = @_;
    my @equations;

    for my $line (split /\n/, $text) {
        next unless $line =~ /\S/;
        my ($target, $nums_str) = split /: /, $line;
        my @nums = split / /, $nums_str;
        push @equations, [$target, \@nums];
    }

    return \@equations;
}

sub evaluate {
    my ($nums, $ops) = @_;
    my $result = $nums->[0];

    for my $i (0 .. $#$ops) {
        my $op = $ops->[$i];
        if ($op eq '+') {
            $result += $nums->[$i + 1];
        } elsif ($op eq '*') {
            $result *= $nums->[$i + 1];
        } elsif ($op eq '||') {
            $result = $result . $nums->[$i + 1];
        }
    }

    return $result;
}

sub generate_operator_combinations {
    my ($operators, $count) = @_;
    return [[]] if $count == 0;

    my @combinations;
    my $subcombinations = generate_operator_combinations($operators, $count - 1);

    for my $op (@$operators) {
        for my $subcomb (@$subcombinations) {
            push @combinations, [$op, @$subcomb];
        }
    }

    return \@combinations;
}

sub can_make_target {
    my ($target, $nums, $operators) = @_;
    my $n_ops = scalar(@$nums) - 1;

    my $combinations = generate_operator_combinations($operators, $n_ops);

    for my $ops (@$combinations) {
        if (evaluate($nums, $ops) == $target) {
            return 1;
        }
    }

    return 0;
}

sub part1 {
    my ($equations) = @_;
    my @operators = ('+', '*');
    my $total = 0;

    for my $eq (@$equations) {
        my ($target, $nums) = @$eq;
        if (can_make_target($target, $nums, \@operators)) {
            $total += $target;
        }
    }

    return $total;
}

sub part2 {
    my ($equations) = @_;
    my @operators = ('+', '*', '||');
    my $total = 0;

    for my $eq (@$equations) {
        my ($target, $nums) = @$eq;
        if (can_make_target($target, $nums, \@operators)) {
            $total += $target;
        }
    }

    return $total;
}

# Main
my $input_file = '../input.txt';
open my $fh, '<', $input_file or die "Cannot open $input_file: $!";
my $text = do { local $/; <$fh> };
close $fh;

my $equations = parse_input($text);

say 'Part 1: ', part1($equations);
say 'Part 2: ', part2($equations);

File: 2024/day07/php/solution.php
#!/usr/bin/env php
<?php

function parseInput($text) {
    $equations = [];
    $lines = array_filter(explode("\n", trim($text)));

    foreach ($lines as $line) {
        list($target, $nums) = explode(': ', $line);
        $numbers = array_map('intval', explode(' ', $nums));
        $equations[] = [(int)$target, $numbers];
    }

    return $equations;
}

function evaluate($nums, $ops) {
    $result = $nums[0];

    for ($i = 0; $i < count($ops); $i++) {
        $op = $ops[$i];

        if ($op === '+') {
            $result += $nums[$i + 1];
        } elseif ($op === '*') {
            $result *= $nums[$i + 1];
        } elseif ($op === '||') {
            $result = (int)($result . $nums[$i + 1]);
        }
    }

    return $result;
}

function generateOperatorCombinations($operators, $count) {
    if ($count === 0) {
        return [[]];
    }

    $result = [];
    $subCombinations = generateOperatorCombinations($operators, $count - 1);

    foreach ($operators as $op) {
        foreach ($subCombinations as $subCombo) {
            $result[] = array_merge([$op], $subCombo);
        }
    }

    return $result;
}

function canMakeTarget($target, $nums, $operators) {
    $nOps = count($nums) - 1;
    $combinations = generateOperatorCombinations($operators, $nOps);

    foreach ($combinations as $ops) {
        if (evaluate($nums, $ops) === $target) {
            return true;
        }
    }

    return false;
}

function part1($equations) {
    $operators = ['+', '*'];
    $total = 0;

    foreach ($equations as list($target, $nums)) {
        if (canMakeTarget($target, $nums, $operators)) {
            $total += $target;
        }
    }

    return $total;
}

function part2($equations) {
    $operators = ['+', '*', '||'];
    $total = 0;

    foreach ($equations as list($target, $nums)) {
        if (canMakeTarget($target, $nums, $operators)) {
            $total += $target;
        }
    }

    return $total;
}

// Main execution
$inputFile = __DIR__ . '/../input.txt';
$text = file_get_contents($inputFile);
$equations = parseInput($text);

echo 'Part 1: ' . part1($equations) . "\n";
echo 'Part 2: ' . part2($equations) . "\n";
